internal/infraenv/infraenv.go

```diff
@@ -31,6 +34,53 @@ func NewManager(log logrus.FieldLogger, db *gorm.DB, objectHandler s3wrapper.API
 	}
 }
 
+func (m Manager) DeleteOrphanInfraEnvs(ctx context.Context, maxDeletePerInterval int, inactiveSince strfmt.DateTime) error {
+	log := logutil.FromContext(ctx, m.log)
+	var infraEnvs []*models.InfraEnv
+	if err := m.db.Limit(maxDeletePerInterval).Where("updated_at < ?", inactiveSince).Find(&infraEnvs).Error; err != nil {
+		return err
+	}
+	for _, infraEnv := range infraEnvs {
+		if infraEnv.ClusterID != "" {
+			_, err := common.GetClusterFromDBWhere(m.db, common.SkipEagerLoading,
+				common.SkipDeletedRecords, "id = ?", infraEnv.ClusterID)
+			if errors.Is(err, gorm.ErrRecordNotFound) {
+				err = m.DeregisterInfraEnvAndHosts(ctx, *infraEnv.ID)
```

Comment:
Isn't this the same logic in either branch? Why look up the cluster at all?

Reply:
If the InfraEnv is related to a cluster, we want to delete it only if the cluster has been deleted.

Reply:
Ah okay, maybe you could deduplicate the `DeregisterInfraEnvAndHosts` call if we return early from the branch where we have a cluster id?

Otherwise maybe add a comment to say we're specifically skipping infraEnvs if they have an existing cluster.

Reply:
Added a comment